File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.vb
Imports System

Namespace Tests.Diagnostics
    Public Interface IFace
        Sub Method6(a As Integer()()) ' Noncompliant {{Make this method private or simplify its parameters to not use multidimensional arrays.}}
'           ^^^^^^^
    End Interface

    Public Class Base
        MustOverride Sub Method5(a As Integer()()) ' Noncompliant
    End Class

    Public Class PublicMethodWithMultidimensionalArray
        Inherits Base
        Implements IFace
        Public Sub Method(a As Integer)
        End Sub
        Public Sub MethodX(a As Integer())
        End Sub

        Public Sub Method1(a As Integer()()) ' Noncompliant
        End Sub
        Sub Method2(a As Integer()()) ' Noncompliant
        End Sub
        Sub Method2b(a As Integer(,)) ' Noncompliant
        End Sub

        Private Sub Method3(a As Integer()())
        End Sub

        Function Method4(a As Integer()()) As Integer ' Noncompliant
            Return 1
        End Function

        Overrides Sub Method5(a As Integer()()) ' Compliant
        End Sub

        Private Sub Method6(a As Integer()()) Implements IFace.Method6 ' Compliant, interface implementation
            Throw New NotImplementedException()
        End Sub
    End Class
    Friend Class Other

        Public Sub Method1(a As Integer()()) ' Compliant
        End Sub

    End Class
End Namespace

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.cs
namespace Tests.Diagnostics
{
    public interface IFace
    {
        void Method5(int[,] a); //Noncompliant
//           ^^^^^^^
    }

    public class Base
    {
        public abstract void Method4(int[,] a); //Noncompliant {{Make this method private or simplify its parameters to not use multidimensional arrays.}}
    }

    public class PublicMethodWithMultidimensionalArray : Base, IFace
    {
        public void Method(int a)
        {
        }
        public void MethodX(int[] a)
        {
        }
        public void Method1(int[][] a) //Noncompliant
        {
        }
        void Method2(int[][] a) //Compliant
        {
        }
        public void Method3(int[,] a) //Noncompliant
        {
        }

        public override void Method4(int[,] a) //Compliant, overrides
        {
        }

        public void Method5(int[,] a) //Compliant, implements interface
        {
        }
    }

    internal class Other
    {
        public void Method1(int[][] a) //Compliant, class is internal
        {
        }
    }
}
